Improve the Quick Tour (#1407)
ballerina-platform/ballerina-dev-website: swan-lake/learn/quick-tour.md

## Installing Ballerina

> 1. [Download](/downloads) Ballerina based on the Operating System you are using. 

Reviewer: ```suggestion
1. [Download](/downloads) Ballerina based on the operating system you are using. 
```

1. [Install](/swan-lake/learn/installing-ballerina) Ballerina. 
1. Set up [Visual Studio Code](/swan-lake/learn/tools-ides/vscode-plugin) for Ballerina.

> ## Creating a Service and a Client to Invoke It

Reviewer: To me this topic is bit confusing. The phrase `a Service and a Client` goes together so the last `it` word is bit confusing.

Author: Changed to "Creating a Service and a Client".

> Let's create a custom Ballerina HTTP service and also an HTTP client to invoke it.

Reviewer 2: Let's remove the "custom" part from here.

Author: Done.

> You can use a Ballerina HTTP client to interact with any HTTP service. Let's create a custom Ballerina HTTP service and also an HTTP client to invoke it.

Reviewer: It seems the sentence `You can use a Ballerina HTTP client to interact with any HTTP service.` is not needed here.

Author: Removed.

> ### Writing a simple 'hello' Service

Reviewer: Use the backtick instead of a single quote.

Author: Back tick styles are not supported by the TOC plugin. Therefore, we have to use single quotes for titles.

Author: This is how it is displayed in the TOC when we use backticks.

<img width="765" alt="Screenshot 2020-11-16 at 13 13 44" src="https://user-images.githubusercontent.com/11707273/99225214-b1f15080-280d-11eb-96d7-464d50aa0e55.png">

Reviewer: What is the issue of using backtick in TOC? I didn't notice the issue in the above image TOC

Author: The word with in the backticks (e.g.,`hello`) doesnt get highlighted in the TOC elements.

## Installing Ballerina

1. [Download](/downloads) Ballerina based on the Operating System you are using. 
1. [Install](/swan-lake/learn/installing-ballerina) Ballerina. 
1. Set up [Visual Studio Code](/swan-lake/learn/tools-ides/vscode-plugin) for Ballerina.

> ## Creating a Client to Invoke a Service 

Reviewer: What is the reason for moving the `Client` example to the top? IMO, services are one of the key interesting feature in ballerina. So shouldn't we start with service first?

Author: Brought "writing the service" part back to the top.

}
```

> ### Running the 'hello' Service

Reviewer: Use the backtick instead of a single quote.  Check all the topics

Author: Back tick styles are not supported by the TOC plugin. Therefore, we have to use single quotes for titles.

[ballerina/http] started HTTP/WS listener 0.0.0.0:9090
```

This means your service is up and running. 

> **Note:** You can test the service by invoking it using an already-available HTTP client. For example, execute the command below in a new CLI tab to use [cURL](https://curl.haxx.se/download.html) as the client.

```bash

> curl http://localhost:9090/hello/sayHello

Reviewer: Output is missing

Author: Added the output.

```

Alternatively, you can create a Ballerina HTTP client and use that to invoke the service as follows.

### Creating a 'helloClient' to Invoke the 'hello' Service

A Ballerina client is a component, which interacts with a network-accessible service. It aggregates one or more actions that can be executed on the network-accessible service and accepts configuration parameters related to the network-accessible service.

There are two kinds of clients in Ballerina, inbound (or ingress) and outbound (or egress) clients. An outbound client object can be used to send messages to a network service.

> Create an outbound Ballerina client as a Ballerina program with a `main` function with the content below to invoke the `hello` service.   

Reviewer: It seems `with the content below` is not appopriate as we are referring to the source code.

Author: Fixed via https://github.com/ballerina-platform/ballerina-dev-website/pull/1407#discussion_r519455090.

> Create an outbound Ballerina client as a Ballerina program with a `main` function with the content below to invoke the `hello` service.   

Reviewer: ```suggestion
Create an outbound Ballerina client as a Ballerina program with a `main` function as follows to invoke the `hello` service.   
```

> **Note**: Returning `error?` allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to handle those errors explicitly.

***hello_client.bal***
```ballerina
import ballerina/http;
import ballerina/io;

public function main() returns @tainted error? {

> 

Reviewer: Please run and validate the below code block. It seems this has invalid syntax for SL versions.

Author: Fixed via https://github.com/ballerina-platform/ballerina-dev-website/pull/1407#discussion_r524003358.

    // Add the relevant endpoint URL to perform the invocation.
    http:Client helloClient = new("http://localhost:9090/hello");

    // Perform a `GET` request to the `hello` service. The remote call 
    // would return an `http:Response` if successful, 
    // or an `error` on failure.
    http:Response helloResp = check helloClient->get("/sayHello");

    // Retrieve the payload as a `string` and print it if the 
    // response of the remote call is successful.
    io:println(check helloResp.getTextPayload());
}
```

### Invoking the 'hello' Service Using the 'helloClient'  

> In a new tab of the CLI, navigate to the location in which you have the `hello_client.bal` file and execute the command below to run the `hello_client.bal` file containing the `main` function (of the client), which invokes the 'helloClient' service.

Reviewer: Use backtick before the `helloClient` .

Author: Done.

### Creating a 'helloClient' to Invoke the 'hello' Service

A Ballerina client is a component, which interacts with a network-accessible service. It aggregates one or more actions that can be executed on the network-accessible service and accepts configuration parameters related to the network-accessible service.

There are two kinds of clients in Ballerina, inbound (or ingress) and outbound (or egress) clients. An outbound client object can be used to send messages to a network service.

Create an outbound Ballerina client as a Ballerina program with a `main` function as follows to invoke the `hello` service.   

> **Note**: Returning `error?` allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to handle those errors explicitly.

***hello_client.bal***
```ballerina
import ballerina/http;
import ballerina/io;

public function main() returns @tainted error? {

    // Add the relevant endpoint URL to perform the invocation.
    http:Client helloClient = new("http://localhost:9090/hello");

>     // Perform a `GET` request to the `hello` service. The remote call 
>     // would return an `http:Response` if successful, 
>     // or an `error` on failure.

Reviewer 3: ```suggestion
    // Perform a `GET` request to the `hello` service. If successful, 
    // the remote call would return an `http:Response` or the payload 
    // (if the `targetType` defaultable parameter is configured).
    // Otherwise an `error` on failure.
```

>     http:Response helloResp = check helloClient->get("/sayHello");

Reviewer 3: ```suggestion
    http:Response helloResp = <http:Response> check helloClient->get("/sayHello");
```

[ballerina/http] started HTTP/WS listener 0.0.0.0:9090
```

This means your service is up and running. 

> **Note:** You can test the service by invoking it using an already-available HTTP client. For example, execute the command below in a new CLI tab to use [cURL](https://curl.haxx.se/download.html) as the client.

```bash
curl http://localhost:9090/hello/sayHello
```

You get the following response.

```bash
Hello Ballerina!
```

Alternatively, you can create a Ballerina HTTP client and use that to invoke the service as follows.

> ### Creating a 'helloClient' to Invoke the 'hello' Service

Reviewer 2: We don't usually call something a "helloClient", similar to hello world programs or a hello service. We can just say "Creating an HTTP Client to...".

Author: Done.

A Ballerina client is a component, which interacts with a network-accessible service. It aggregates one or more actions that can be executed on the network-accessible service and accepts configuration parameters related to the network-accessible service.

There are two kinds of clients in Ballerina, inbound (or ingress) and outbound (or egress) clients. An outbound client object can be used to send messages to a network service.

> Create an outbound Ballerina client as a Ballerina program with a `main` function as follows to invoke the `hello` service.   

Reviewer 2: Why are we using the term "outbound" here? Are there inbound Ballerina clients as well? If there's no specific reason, let's remove this.

Author: Done.

> > **Note**: Returning `error?` allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to handle those errors explicitly.

Reviewer: ```suggestion
> **Note**: Returning `error?` from the `main` function allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to handle those errors explicitly.
```

***hello_client.bal***
```ballerina
import ballerina/http;
import ballerina/io;

public function main() returns @tainted error? {

> 

Reviewer 2: Remove the extra line space here.

Author: Done.

    // Add the relevant endpoint URL to perform the invocation.
    http:Client helloClient = new("http://localhost:9090/hello");

    // Perform a `GET` request to the `hello` service. If successful, 
    // the remote call would return an `http:Response` or the payload 
    // (if the `targetType` defaultable parameter is configured).
    // Otherwise an `error` on failure.
    http:Response helloResp = <http:Response> check helloClient->get("/sayHello");

    // Retrieve the payload as a `string` and print it if the 
    // response of the remote call is successful.
    io:println(check helloResp.getTextPayload());
}
```

> ### Invoking the 'hello' Service Using the 'helloClient'  

Reviewer 2: Let's remove the "helloClient" terms as mentioned above.

Author: Done.

> In a new tab of the CLI, navigate to the location in which you have the `hello_client.bal` file and execute the command below to run the `hello_client.bal` file containing the `main` function (of the client), which invokes the `helloClient` service.

Reviewer: ```suggestion
In a new tab of the CLI, navigate to the location in which you have the `hello_client.bal` file and execute the command below to run the `hello_client.bal` file containing the `main` function (of the client), which invokes the `hello` service.
```

[ballerina/http] started HTTP/WS listener 0.0.0.0:9090
```

This means your service is up and running. 

> **Note:** You can test the service by invoking it using an already-available HTTP client. For example, execute the command below in a new CLI tab to use [cURL](https://curl.haxx.se/download.html) as the client.

```bash
curl http://localhost:9090/hello/sayHello
```

Alternatively, you can create a Ballerina HTTP client and use that to invoke the service as follows.

### Creating a 'helloClient' to Invoke the 'hello' Service

There are two kinds of clients in Ballerina, inbound (or ingress) and outbound (or egress) clients. An outbound client object can be used to send messages to a network service.

Create an outbound Ballerina client as a Ballerina program with a `main` function with the content below to invoke the `hello` service.   

> **Note**: Returning `error?` allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to handle those errors explicitly.

> ***hello_client.bal***

Reviewer: Do we really want another client creation example in quick tour? Having one example (`Sunrise API Client`) is enough IMO.

Author: Removed the Sunrise API related use case. Now, we have a simple and easy-to-follow structure and flow explaining the use case in a straightforward manner.
